Add algorithm to load gauss cube files

RichardWaiteSTFC opened the pull request:
**Report to:** Matthias Gutmann (SXD)


**Summary of work**
Loader for Gaussian cube format (file format used to store volumetric data) - but data into a 3d MDHisto workspace.
This file format is used by SXD to store results of diffuse scattering simulations

**To test:**

(1) Run this script (same as unit test - so make sure you build test data)
```
ws = LoadGaussCube(Filename="gauss_cube_example.cube", Names=['[H,0,0]','[0,K,0]','[0,0,L]'],
                          Frames='HKL,HKL,HKL',
                          Units='rlu,rlu,rlu')
```
(2) Check you get an error if there are not 3 elements in the `Names`, `Frames` and `Units` input parameters

Fixes #35702 
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

RichardWaiteSTFC committed:
Add biolerplate for new algorithm LoadGaussCube

RichardWaiteSTFC committed:
Code to parse file and call CreateMDHistoWorkspace

RichardWaiteSTFC committed:
Fix bug skipping first two rows of data

RichardWaiteSTFC committed:
Add input validation to ensure 3D

RichardWaiteSTFC committed:
Add unit test

RichardWaiteSTFC committed:
Fix cppcheck (str assigned value not used)

RichardWaiteSTFC committed:
Fix new test

RichardWaiteSTFC committed:
Add doc page and release notes

RichardWaiteSTFC committed:
Fix gcc warnings

AnthonyLim23 commented on Framework/MDAlgorithms/inc/MantidMDAlgorithms/LoadGaussCube.h:
```diff
@@ -0,0 +1,31 @@
+namespace Mantid {
+namespace MDAlgorithms {
+
+/** LoadGaussCube : TODO: DESCRIPTION
```
Did you mean to leave this to do in?

RichardWaiteSTFC replied:
Nope, will remove!

AnthonyLim23 commented on Framework/MDAlgorithms/src/LoadGaussCube.cpp:
```diff
@@ -0,0 +1,145 @@
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string LoadGaussCube::summary() const {
+  return "Algorithm to load gauss cube files and output an 3D MDHistoWorkspace.";
```
```suggestion
  return "Algorithm to load gauss cube files and output a 3D MDHistoWorkspace.";
```

AnthonyLim23 commented on Framework/MDAlgorithms/src/LoadGaussCube.cpp:
```diff
@@ -0,0 +1,145 @@
+  // open file
+  std::ifstream in(Filename.c_str());
+  // skip header
+  for (int iline = 0; iline < 2; ++iline) {
```
what is `iline`?

RichardWaiteSTFC replied:
index of line

AnthonyLim23 replied:
I would expect it to be either `indexOfLine` or `line_i` (I personally like the second one better, but thats me with my science hat on)

AnthonyLim23 commented on Framework/MDAlgorithms/src/LoadGaussCube.cpp:
```diff
@@ -0,0 +1,145 @@
+  std::vector<double> extents(6, 0.0);
+  getWord(in, false); // ignore the first element of this row
+  std::string str;
+  for (size_t idim = 0; idim < 3; ++idim) {
```
idim?

RichardWaiteSTFC replied:
index of dimension

AnthonyLim23 commented on Framework/MDAlgorithms/test/LoadGaussCubeTest.h:
```diff
@@ -0,0 +1,61 @@
+    TS_ASSERT_THROWS_NOTHING(alg.initialize())
+    TS_ASSERT(alg.isInitialized())
+  }
+
```
can we have a test to check the validation?

RichardWaiteSTFC replied:
Yep good shout!

AnthonyLim23 commented on docs/source/algorithms/LoadGaussCube-v1.rst:
```diff
@@ -0,0 +1,31 @@
+
+    from mantid.simpleapi import *
+
+    ws = LoadGaussCube(Filename="gauss_cube_example.cube", Names=['[H,0,0]','[0,K,0]','[0,0,L]'], Frames='HKL,HKL,HKL', Units='rlu,rlu,rlu')
```
can we not add a doc test here? If we print something we can say what the expected output is.

RichardWaiteSTFC replied:
Can do, but given the output is a single workspace, I'm not sure what that would add over the unit tests?

AnthonyLim23 replied:
Just an extra layer of testing for us. It also tells users what sort of result to expect, you can use it to highlight key differences to "normal" output. I always feel happier seeing a test in the docs as I can check if it works as the docs expect

AnthonyLim23 commented:
I cannot use it from the algorithm dialog 
![image](https://github.com/mantidproject/mantid/assets/25006392/73bad18b-c3ac-447f-a764-9480e082465f)

RichardWaiteSTFC commented:
> I cannot use it from the algorithm dialog ![image](https://user-images.githubusercontent.com/25006392/253532776-73bad18b-c3ac-447f-a764-9480e082465f.png)

As discussed offline this is due to the way strings of lists are parsed. In this case we want to ignore the `,` in e.g. `[H,0,0]` but not the `,` that delimits the different axes names - to do this in a script you put the names in a python list, but you cannot put it in a single string as required in the algorithm dialog (note you need to remove the speech quotes as it is already interpreted as a string).
It was a known issue with `CreateMDHistoWorkspace` but there is a workaround for this use-case that I can add try to use in this PR
https://github.com/mantidproject/mantid/blob/c7d4d55d719739e9e2f029ee88c62f9991f2d577/Framework/MDAlgorithms/src/ImportMDHistoWorkspaceBase.cpp#L203

RichardWaiteSTFC committed:
Use parseNames in ImportMDHistoWorkspaceBase to accept comma in label

E.g. accept labels of form [H,0,0] - see updated names in test

RichardWaiteSTFC committed:
Add test for input validation forcing 3 Names, Units, Frames

The above properties have to be of equal length (validation in CreateMDHistoWorkspace) so don't need to test each one separately

RichardWaiteSTFC committed:
Add doc test

RichardWaiteSTFC committed:
Rename variables as per PR review comments

AnthonyLim23 reviewed (approved):
It works as expected, even from the algorithm dialog. The code is also very readable now. :shipit:
